app/src/main/java/com/google/android/samples/socialite/ui/player/preloadmanager/PreloadManagerWrapper.kt

```diff
@@ -0,0 +1,128 @@
+            context: Context,
+        ): PreloadManagerWrapper {
+            val trackSelector = DefaultTrackSelector(context)
+            trackSelector.init({}, DefaultBandwidthMeter.getSingletonInstance(context))
+            val renderersFactory = DefaultRenderersFactory(context)
+            val preloadManager = DefaultPreloadManager(
+                PreloadStatusControl(),
+                DefaultMediaSourceFactory(context),
+                trackSelector,
+                DefaultBandwidthMeter.getSingletonInstance(context),
+                DefaultRendererCapabilitiesList.Factory(renderersFactory),
+                loadControl.allocator,
+                playbackLooper,
+            )
+            return PreloadManagerWrapper(preloadManager)
+        }
+    }
+
+    /** Add initial list of videos to the preload manager. */
+    fun init(mediaList: List<TimelineMediaItem>) {
+        for ((index, item) in mediaList.withIndex()) {
+            if (item.type == TimelineMediaType.VIDEO) {
+                addMediaItem((MediaItem.fromUri(item.uri)), index)
+            }
+        }
+    }
+
+    /** Sets the index of the current playing media. */
+    fun setCurrentPlayingIndex(currentPlayingIndex: Int) {
+        defaultPreloadManager.setCurrentPlayingIndex(currentPlayingIndex)
+    }
+
+    /** Sets the size of the Preload Queue. */
+    fun setPreloadWindowSize(preloadWindowSize1: Int) {
+        preloadWindowSize = preloadWindowSize1
+    }
+
+    /** Add media item to the preload manager */
+    fun addMediaItem(mediaItem: MediaItem, index: Int): Boolean {
+        // item not found in list. This could happen if the list is auto purged or contents refreshed by APIs
+        if (index < 0) {
+            return false
+        }
+        // item already added , avoid duplicates
+        if (preloadWindow.contains(Pair(mediaItem, index))) {
+            return false
+        }
+
+        // Window full, purge old data added at the start of the queue
+        if (preloadWindow.size >= preloadWindowSize) {
+            defaultPreloadManager.remove(preloadWindow.first().first)
+            preloadWindow.removeFirstOrNull()
+        }
+        // Add video to preload list
+        preloadWindow.add(Pair(mediaItem, index))
+        defaultPreloadManager.add(mediaItem, index)
+        defaultPreloadManager.invalidate()
+        return true
+    }
+
+    /** Releases the preload manager. This must be called on the main thread */
+    @MainThread
+    fun release() {
+        defaultPreloadManager.release()
+    }
+
+    /** Customize time to preload, by default as per ranking data */
+    @androidx.media3.common.util.UnstableApi
+    class PreloadStatusControl : TargetPreloadStatusControl<Int> {
+        override fun getTargetPreloadStatus(rankingData: Int): DefaultPreloadManager.Status {
+            // By default preload first 5 seconds of the video
+            return DefaultPreloadManager.Status(STAGE_LOADED_TO_POSITION_MS, 5000L)
```

Comment:
Optional: We can try only preloading the most adjacent video for the first 5 seconds, and preload the rest of videos less (eg. 2 seconds) to save memory allocations. Or even more reduction: most adjacent videos - 5 seconds, videos one step further - 2 seconds, videos two steps further - 1 second.

This will also save some memory allocations for playback. On the other hand, we can try if this reduced version will hurt the key metrics of starting latency, if not, it's a good deal.